pkg/certificate/providers/config.go

```diff
@@ -275,6 +274,81 @@ func (c *MRCProviderGenerator) getCertManagerOSMCertificateManager(mrc *v1alpha2
 	if err != nil {
 		return nil, fmt.Errorf("error instantiating Jetstack cert-manager client: %w", err)
 	}
-
 	return cmClient, nil
 }
+
+// getCertFromKubernetes is a helper function that loads a certificate from a models.Secret
+func getCertFromKubernetes(ns string, secretName string, computeClient compute.Interface) (*certificate.Certificate, error) {
+	certSecret := computeClient.GetSecret(secretName, ns)
+	if certSecret == nil {
+		// TODO(#3962): metric might not be scraped before process restart resulting from this error
+		log.Error().Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrFetchingCertSecret)).
+			Msgf("Could not retrieve certificate secret %q from namespace %q", secretName, ns)
+		return nil, certificate.ErrSecretNotFound
+	}
+	pemCert, ok := certSecret.Data[constants.KubernetesOpaqueSecretCAKey]
+	if !ok {
+		// TODO(#3962): metric might not be scraped before process restart resulting from this error
+		log.Error().Err(certificate.ErrInvalidCertSecret).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrObtainingCertFromSecret)).
+			Msgf("Opaque k8s secret %s/%s does not have required field %q", ns, secretName, constants.KubernetesOpaqueSecretCAKey)
+		return nil, certificate.ErrInvalidCertSecret
+	}
+
+	pemKey, ok := certSecret.Data[constants.KubernetesOpaqueSecretRootPrivateKeyKey]
+	if !ok {
+		// TODO(#3962): metric might not be scraped before process restart resulting from this error
+		log.Error().Err(certificate.ErrInvalidCertSecret).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrObtainingPrivateKeyFromSecret)).
+			Msgf("Opaque k8s secret %s/%s does not have required field %q", ns, secretName, constants.KubernetesOpaqueSecretRootPrivateKeyKey)
+		return nil, certificate.ErrInvalidCertSecret
+	}
+
+	cert, err := certificate.NewCertificateFromPEM(pemCert, pemKey, pemCert, "", "")
+	if err != nil {
+		log.Error().Err(err).Msg("Failed to create new Certificate from PEM")
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+// getCertificateFromSecret is a helper function that ensures creation and synchronization of a certificate
+// using Kubernetes Secrets backend and API atomicity.
+func getCertificateFromSecret(ns string, secretName string, cert *certificate.Certificate, computeClient compute.Interface) (*certificate.Certificate, error) {
```

Comment:
let's rename this to something more indicative of what it does. how about `ensureSecretExists`

let's also make this a pointer receiver on the MRCProviderGenerator, so we don't have to pass a compute client interface around